<h1 id="bowtie2">bowtie2</h1>
<p>From the manual: <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml"><em>"Bowtie 2 is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences"</em></a>.</p>
<p><code>bowtie2</code> can be used to:</p>
<ul>
<li>index reference FASTA nucleotide genomes/sequences</li>
<li>align FASTQ sequencing reads to those genomes/sequences</li>
</ul>
<h2 id="differences-between-bowtie-and-bowtie2">Differences between <code>bowtie</code> and <code>bowtie2</code></h2>
<ul>
<li><code>bowtie2</code> has no upper limit on read length</li>
<li><code>bowtie2</code> can make gapped alignments</li>
<li><code>bowtie2</code> is more flexible for paired-end alignment</li>
<li><code>bowtie2</code> is faster and more memory efficient</li>
<li><code>bowtie</code> is advantageous over <code>bowtie2</code> for relatively short sequencing reads (50bp or less)</li>
</ul>
<h2 id="indexing-a-reference-genomesequence-using-bowtie2-build">Indexing a reference genome/sequence using <code>bowtie2-build</code></h2>
<p>Before aligning reads to a reference genome with <code>bowtie2</code>, it must be indexed using <code>bowtie2-build</code>. This command will create six files with the extensions <code>.1.bt2</code>, <code>.2.bt2</code>, <code>.3.bt2</code>, <code>.4.bt2</code>, <code>.rev.1.bt2</code>, and <code>.rev.2.bt2</code>. These six files together are the index. Once an index has been created, the original reference genome/sequence is no longer needed to align reads. Here's an example <code>bowtie2-build</code> command:</p>
<pre><code>$ bowtie2-build reference_sequence.fasta index_name
</code></pre>
<p>In this command, the <code>reference_sequence.FASTA</code> is the nucleotide FASTA sequence we want to index, and <code>index_name</code> is the name of the index. There will be six files beginning with the <code>index_name</code> in the output directory: <code>index_name.1.bt2</code>, <code>index_name.2.bt2</code>, <code>index_name.3.bt2</code>, <code>index_name.4.bt2</code>, <code>index_name.rev.1.bt2</code>, and <code>index_name.rev.2.bt2</code>. There's no need to specify any of these files individually, just the <code>index_name</code> alone is enough to refer to the entire index.</p>
<h2 id="aligning-reads-to-an-indexed-genomesequence-using-bowtie2">Aligning reads to an indexed genome/sequence using <code>bowtie2</code></h2>
<p>Now that the genome has been indexed, FASTQ sequencing reads can be aligned to it. This is done using the <code>bowtie2</code> command. Here's an example <code>bowtie2</code> command:</p>
<pre><code>$ bowtie2 --no-unal -p n -x index_name -1 reads_1.fastq -2 reads_2.fastq -S output.sam
</code></pre>
<p>In this command...</p>
<ol>
<li><strong><code>--no-unal</code></strong> is an optional argument, meaning reads that do not align to the reference genome will not be written to <code>sam</code> output</li>
<li><strong><code>-p</code></strong> is the number (<em>n</em>) of processors/threads used</li>
<li><strong><code>-x</code></strong> is the genome index</li>
<li><strong><code>-1</code></strong> is the file(s) containing mate 1 reads</li>
<li><strong><code>-2</code></strong> is the file(s) containing mate 2 reads</li>
<li><strong><code>-S</code></strong> is the output alignment in <code>sam</code> format</li>
</ol>
<h2 id="demonstration">Demonstration</h2>
<p>In this video, <code>bowtie2-build</code> is used to index <code>example_nucleotide_sequence.fasta</code>, and the command <code>bowtie2</code> is used to align reads to this bowtie2 index.</p>
<script id="asciicast-306546" src="https://asciinema.org/a/306546.js" async></script>
<h2 id="further-reading">Further reading</h2>
<ol>
<li>The <code>bowtie2</code> manual: <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml</a></li>
</ol>
